[wip/hyperstart] Initial proxy implementation

dlespiau opened the pull request:
This PR is for a new wip/hyperstart topic branch. I took the liberty to create the branch, but review comments are more than welcomed, of course! hence this PR.

This initial implementation is definitely not complete but is enough for a client to open the proxy socket, register a qemu instance and issue hyperstart commands. Documentation should cover that (proxy/README.md).

There is definitely more work to be done The main part is having a way to hand off an I/O socket to the runtime/shim and forward the data from the hyperstart I/O channel to shim. Current line of thinking around that is in https://github.com/01org/cc-oci-runtime/issues/303

dlespiau commented:
New version with:
  - a new `bye` command for the runtime to undo the `hello`
  - a couple of minor cleanups squashed in (comment, jsonproto using json.RawMessage and not a pointer to it, ...)

Unit tests are still missing, but might as well  describe how I test so far:
  - Use the daemon hyperstart branch at https://github.com/dlespiau/hyperstart/tree/20160928-daemon
  - That branch has a commit that will make hyperstart use a couple of sockets instead of the serial links so one can run hyperstart on the host (https://github.com/dlespiau/hyperstart/commit/f50f9fcb327fca8a61190859f8fb52f008b6f685, use the `--enable-daemon` configure option)
  - I have files with a succession of commands like:

```
$ cat ping
{ "id": 0, "command": "hello", "data": { "containerId": "foo", "ctlSerial": "/tmp/sh.hyper.channel.0.sock", "ioSerial": "/tmp/sh.hyper.channel.1.sock"  } }
{ "id": 1, "command": "hyper", "data": { "name": "ping" }}
{ "id": 2, "command": "bye"}
```

and then:

```
$ socat /var/run/cc-oci-runtime/proxy.sock  - < ping
{"id":0,"success":true}
{"id":1,"success":true}
{"id":2,"success":true}
```

sameo commented on proxy/README.md:
```diff
@@ -0,0 +1,99 @@
+
+This socket can be found at: `${localstatesdir}/run/cc-oci-runtime/proxy.sock`.
+
+## Protocol
```
You may want to extensively comment the prototype implementation and turn the generated godoc into a spec of some sort. But at some point this protocol needs to be formalized a bit further.

sameo commented on proxy/README.md:
````diff
@@ -0,0 +1,99 @@
+
+```
+type Request struct {
+	Id      uint             `json:"id"`
````
Is `Id` really needed here ? For a given client the protocol is synchronous so you would send the response back to the last received command, in order.

dlespiau replied:
Not it's not really needed, I thought about removing it but then just forgot.

jodh-intel replied:
+1 for removing `Id`.

sameo commented on proxy/README.md:
````diff
@@ -0,0 +1,99 @@
+```
+type Request struct {
+	Id      uint             `json:"id"`
+	Command string           `json:"command"`
````
I'd call that one ID and not Command. Here you have commands, hyper commands, requests and responses. I find it a bit confusing.

sameo commented on proxy/README.md:
````diff
@@ -0,0 +1,99 @@
+}
+```
+
+The `hello` command is issued first after connecting to the proxy socket.  It's
````
So it's going to sound like nitpicking, but to me `hello` is not a command but a command payload. The command is the request and it contains a command/ID string.

sameo commented on proxy/README.md:
````diff
@@ -0,0 +1,99 @@
+
+```
+type hyperCmd struct {
+        Name string          `json:name`
````
Just to be less confusing again, we might want to call that one HyperName.

sameo commented on proxy/jsonproto.go:
```diff
@@ -0,0 +1,138 @@
+			Error:   err.Error(),
+			Data:    respMap,
+		}
+	} else if respMap != nil {
```
Wouldn't:

```
else {
           resp = &Response{
			Id:      req.Id,
			Success: true,
			Data:    respMap,
}
```

be enough ?

dlespiau replied:
looks like it would be!

sameo commented on proxy/jsonproto.go:
```diff
@@ -0,0 +1,138 @@
+	Data    map[string]interface{} `json:"data,omitempty"`
+}
+
+// XXX: could do with its own package to remove that ugly namespacing
```
Indeed :) And it could be called protocol.go. The JSON formatting is only an implementation detail.

sameo commented on proxy/proxy.go:
```diff
@@ -0,0 +1,166 @@
+		go proto.Serve(conn, newClient)
+	}
+}
+
```
This is missing a `Stop()` API where you would for example notify all the client routines that they should exit, and wait for them to be done.

dlespiau replied:
A global `stop()` doesn't seem like it'd fit the model. Each client has (now) a hello()/bye(), but globally noone can tell to proxy to "stop()" because they only have local knowledge of their own qemu lifetime.

What would be possible though, is to exit the proxy once we've lost all clients (or even do that after a time without new clients), a new client would start the proxy again through socket activation.

sameo replied:
If I kill the proxy, all its go routines will leak. So a `defer Stop()` or anything that would clean the proxy resources would be nice to have.

dlespiau commented on proxy/jsonproto.go:
```diff
@@ -0,0 +1,138 @@
+	ctx := &clientCtx{
+		conn:     conn,
+		userData: userData,
+		reader:   bufio.NewReader(conn),
```
I liked the new line separator between requests, but it shouldn't be needed if we just use json.NewDecoder() and it's true that line oriented messages are more CPU work as well (to find the '\n' in the input data). Will change.

jodh-intel commented on .travis-setup.sh:
```diff
@@ -22,6 +22,23 @@ set -e -x
+#
+
+go_tarball=go${go_version}.linux-amd64.tar.gz
+curl -L -O https://storage.googleapis.com/golang/$go_tarball
```
I'd always quote strings like this to ensure they are future-proof against weird expansions.

jodh-intel commented on .travis-setup.sh:
```diff
@@ -22,6 +22,23 @@ set -e -x
+# Unfortunately, go doesn't support vendoring outside of GOPATH (maybe in 1.8?)
+# So, we setup a GOPATH tree with our vendored dependencies.
+# See: https://github.com/golang/go/issues/14566
+mkdir -p $GOPATH/src
```
same comment here.

jodh-intel commented on .travis.yml:
```diff
@@ -20,6 +22,7 @@ env:
     - coverity_scan_run_condition='"$CC" = gcc'
 
 before_script:
+- env
```
why?

dlespiau replied:
left over from debugging :(

jodh-intel commented on Makefile.am:
```diff
@@ -61,6 +61,8 @@ GENERATED_FILES = \
 		-e 's|[@]bindir[@]|$(bindir)|g' \
+		-e 's|\@libexecdir\@|$(libexecdir)|' \
+		-e "s|\@localstatedir\@|$(localstatedir)|" \
 		-e 's|[@]BUNDLE_TEST_PATH[@]|$(BUNDLE_TEST_PATH)|g' \
```
These 2 lines are for some reason inconsistent with the existing code.

jodh-intel commented on Makefile.am:
```diff
@@ -61,6 +61,8 @@ GENERATED_FILES = \
 	$(AM_V_GEN)$(SED) \
 		-e 's|[@]bindir[@]|$(bindir)|g' \
+		-e 's|\@libexecdir\@|$(libexecdir)|' \
+		-e "s|\@localstatedir\@|$(localstatedir)|" \
```
These 2 lines are for some reason inconsistent with the existing code.

jodh-intel commented on Makefile.am:
```diff
@@ -194,6 +196,41 @@ cc_oci_runtime_CFLAGS = \
+systemdservice_in_files =		\
+	proxy/cc-proxy.service.in	\
+	proxy/cc-proxy.socket.in	\
+	$(NULL)
```
I guess this is to avoid the hassle of deciding whether to add a backslash each time the list is updated? If so, I'd still define `NULL` to be nothing to make it clearer (even better, with a comment explaining the intent ;-)

dlespiau replied:
Removed it. If we need to document what it does because it's not clear enough as is, might as well not use it at all :)

jodh-intel commented on configure.ac:
```diff
@@ -43,6 +43,9 @@ AM_PROG_CC_C_O
 AC_PROG_MKDIR_P
 
+# We need go!
+AC_CHECK_PROG(GO,[go],[go],[no])
```
That line is classic ;)

jodh-intel commented on proxy/COPYING:
```diff
@@ -0,0 +1,202 @@
+
+                                 Apache License
```
Why is this code licensed differently to the rest? When was this decided?

jodh-intel commented on proxy/README.md:
```diff
@@ -0,0 +1,105 @@
+
+The `bye` commands does the opposite of what `hello` does, indicating to the
+proxy it should release resources created by `hello`. This command has no
+parameter.
```
So.... what happens if a client calls `hello` then dies? Presumably the proxy can never release the associated resources?

jodh-intel commented on proxy/README.md:
````diff
@@ -0,0 +1,105 @@
+```
+
+The proxy can output log messages on stderr, which are automatically
+handled by systemd and ca be viewed with:
````
s/ca/can/

jodh-intel commented on proxy/cc-proxy.service.in:
```diff
@@ -0,0 +1,8 @@
+[Unit]
+Description=Clear Containers Proxy
```
Might be useful to add:


```
[Unit]
Documentation=https://github.com/01org/cc-oci-runtime/
```

... to these units.

jodh-intel commented on proxy/jsonproto.go:
```diff
@@ -0,0 +1,137 @@
+		if err == io.EOF {
+			return
+		} else if err != nil {
+			fmt.Fprintf(os.Stderr, "read error: %s", err)
```
Might be safer to standardise on `%v`.

jodh-intel commented on proxy/proxy.go:
```diff
@@ -0,0 +1,186 @@
+	if _, ok := proxy.vms[hello.ContainerId]; ok {
+
+		proxy.Unlock()
+		return nil, errors.New("container already registered")
```
include `ContainerId` in error to help problem determination maybe?

jodh-intel commented on proxy/proxy.go:
```diff
@@ -0,0 +1,186 @@
+	fds := listenFds()
+
+	if len(fds) != 1 {
+		return fmt.Errorf("couldn't find activated socket (%d)", len(fds))
```
The error message doesn't explain what the value in brackets is.

jodh-intel commented on proxy/proxy.go:
```diff
@@ -0,0 +1,186 @@
+	fd := fds[0]
+	l, err := net.FileListener(fd)
+	if err != nil {
+		return fmt.Errorf("couldn't listen on socket: %s", err.Error())
```
Log `fd`?

jodh-intel commented on proxy/vm.go:
```diff
@@ -0,0 +1,237 @@
+	needRead := 8
+	length := 0
+	read := 0
+	buf := make([]byte, 512)
```
I'd convert those repeated magic numbers to `const`'s.

jodh-intel commented on proxy/vm.go:
```diff
@@ -0,0 +1,237 @@
+	binary.BigEndian.PutUint32(msg[4:], uint32(length))
+	copy(msg[8:], m.Message)
+
+	_, err = vm.ctl.Write(msg)
```
No check on `err`.

jodh-intel commented on proxy/vm.go:
```diff
@@ -0,0 +1,237 @@
+	_, err = vm.ctl.Write(msg)
+
+	// Consume the ack (NEXT) message
+	next, err := vm.readMessage()
```
No check on `err`.

jodh-intel commented on proxy/vm.go:
```diff
@@ -0,0 +1,237 @@
+	if next.Code != hyper.INIT_NEXT {
+		return errors.New("didn't receive NEXT from hyperstart")
+	}
+	if len(next.Message) != 4 {
```
Another magic - can you document this function (or point to reference URL) and use `const`'s?

devimc commented on Makefile.am:
```diff
@@ -194,6 +196,41 @@ cc_oci_runtime_CFLAGS = \
+systemdservice_files = 		\
+	proxy/cc-proxy.service	\
+	proxy/cc-proxy.socket	\
+	$(NULL)
```
Hi @dlespiau 
did you forget to create symlinks to $(systemdsystemunitdir) ?

dlespiau replied:
Hum, not sure I follow you, why would I create symlinks? The above installs the units file into $(systemdsystemunitdir)

devimc replied:
uhmm, I didn't know that
I did something different in other project https://github.com/clearlinux/micro-config-drive/blob/master/Makefile.am#L77

dlespiau replied:
Oh but that one is a different problem. The symlinks in the  multi-user.target.wants directory control which services are started at boot time. That decision doesn't belong to the "make install" time, but to the admin/user of that machine or to the packaging system (deciding which services should be started by default in a distribution) . Same thing for sockets.target.wants.

The way to control that is through `systemctl enable`. For us that will be

```
$ sudo systemctl enable cc-proxy
Created symlink from /etc/systemd/system/multi-user.target.wants/cc-proxy.service to /usr/lib/systemd/system/cc-proxy.service.
```

and

```
$ sudo systemctl enable cc-proxy.socket
Created symlink from /etc/systemd/system/sockets.target.wants/cc-proxy.socket to /usr/lib/systemd/system/cc-proxy.socket.
```

dlespiau commented:
Changelog of the last push:

### New feature

- Allow the proxy to start without being socket activated for debugging purposes.

### Review comments
  
- Renamed jsonproto.go to protocol.go and removed the Json prefix to the various types
- Eliminated a branch in the response serialization
- Used a json.Decoder instead of splitting lines and unmarshalling each line to
  JSON objects
- Added more details about the cc-proxy place in the big picture and its role
- Talk about "payloads" when describing the content of requests
- Split the API in its own package so we can generate a nice godoc page from it
  and keep the protocol documention inline with the code
    https://godoc.org/github.com/dlespiau/cc-oci-runtime/proxy/api
- Renamed "Name" to "HyperName" in the hyper payload
- Removed the Id field in Requests and renamed "Name" to "Id"
- Mass addition of shell quotes, just in case
- Removed spurious env in travis.yml
- Made the sed replacement commands look the same as the ones already there
- Fixed s/can/can/ typo in the README file
- Added a Documentation tag to systemd's service files
- Use %v to print Errors
- Various little improvements in errors messages
- Various little improvements error checking

###  Not fully addressed

- Interaction between SIGTERM and goroutines. I did quite a bit of testing and
  SIGTERM terminates cc-proxy fine, even with a client still connected.

- Robustness. In a more general sense of the above, there are many points of
  failures: any part of the chain can fail (runtime, shim, proxy, qemu,
  hyperstart) and the all those corner cases will need work to ensure we have a
  way to recover.
        
  There're not only the failure cases but also what to do if we want to support
  upgrading each element, with potentially a new version of the protocol.

  All of the above is non trivial and needs to be thoroughly tested. Given that
  we don't even have enough protocol to get things running and nothing is
  tested at all so far, I think I'd rather have the basics working and covered
  before we look at the end-to-end robustness.

- Not all "magic" values are gone, all of them are in read/writeMessage() with
  code from runv.

dlespiau commented:
#### New updates:

- Added the testify/assert framework for unit tests
- protocol.go is now somewhat tested
- As a result, cleaned up the main loop of the JSON protocol logic to return proper error messages to clients or close the connection when a client is misbehaving.

jcvenegas commented on proxy/README.md:
```diff
@@ -0,0 +1,91 @@
+- `cc-proxy`'s main role is to:
+    - Arbitrate access to the `hyperstart` control channel between all the
+      instances of the OCI runtimes and `cc-shim`.
+    - Route the I/O streams between the various shim instances and `hyperstart`.
```
So the idea is that the cc-shim will read IO streams and shim will be responsible to write to the correct place the tty output. As well that shim will take container input and ask to the proxy to send it to hyperstart -> process

dlespiau commented:
New update! almost there, just allocateIO + I/O stream forwarding to finish debugging/testing

### Featured

- Added a `allocateIO` payload for clients to ask seq numbers for hyperstart
  execcmd
- Added an `attach` payload for clients to attach themselves to an already
  registered VM (registered with `hello`). This is to use with start, exec, delete to link a new connection to the right qemu/hyperstart.
- Added a client go API wrapping the JSON RPC, useful to write proxy clients in
  go
- Improved test coverage to ~50%
- Big missing piece is testing the io forwarding (and actually the
  implementation is missing a piece of code as well)

### Details

- Added the start of a mocking library (tests/mock) to build test harnesses
  around our various components. There's only hyperstart there for now.
- Covered the "easy" proxy commands: hello, attach, hyper, bye
- Fixed the logic receiving the acks from the control channel (Julio reported
  it and it's now covered by unit testing)
- Added logic to route I/O streams both ways between clients and hyperstart
  (not tested yet!)
- Fixed Socketpair() used in tests to close duped fds
- Add a test to make sure we close the connection if a client start sending
  malformed requests
- More generally, close the connection when encountering non-recoverable errors
  (ig, couldn't write the response back to the client for whatever reason)
- Added facilities to pass fds over AF_UNIX sockets
- Arbitrate access to the ctl channel so we don't have mutiple clients
  interleaving read/writes on that channel
- Changed socket permission to 666 (from 777) in the non-socket activated case
- Fixed client<->vm association after a `Bye` (found by unit/interface tests!)
- Fixed the json properties on the api structures (missing "")
- Added patch from Julio to give an example of "hyper" payload, "startpod",
  that has data instead of the "ping" hyper message

amshinde commented on proxy/api/client.go:
```diff
@@ -0,0 +1,163 @@
+		CtlSerial:   ctlSerial,
+		IoSerial:    ioSerial,
+	}
+
```
What if we add an nstreams variable for the Hello command? We can get away with executing the AllocateIO command for the create-start sequence.

dlespiau replied:
Right I thought about that. But then allocateIo is needed for two things:

  - start/create (hello + allocateIo)
  - exec (attach + allocateIo)

I've separated the initial workload that does an "hello" from exec workload that only do an "attach". Because of that, I added allocateIo as an orthogonal command that can be done in either case.

We could still decide to compress the messages in one (hello and attach could both ask to allocate the streams), may have some boot time gains.

amshinde replied:
Yes, I was thinking along the same lines - combining hello into (hello +allocateIO ) for the create-start sequence and attach into (attach + allocateIO) for the exec sequence to get some performance gain.

dlespiau commented:
### Featured

- (**OOPS**) changed the client <-> proxy protocol: every request/response is now
  prefixed a header with the size of that request/response (README updated with
  details)
- Added basic tests for the I/O forwarding between the I/O fd (the fd sent
  after the allocateIO response) and hyperstart. Seems to even work somewhat!
 
### Details

- Fixed an ordering issue in the `Bye` and `Hello` tests that was causing them to
  fail intermittently
- Set the VM sockets (the connections to hyperstart in vm.go) back to nil after
  having waited for the goroutine using them to be finished instead of just
  after closing the sockets. Otherwise the goroutine can panic with null
  derefences (found by the unit tests!)
- The files backing up the test sockets in the mock hyperstart are now unlinked
  on Close()
- Fixed an assertion that was a bit too aggressive in the mock hyperstart
  (asserting that Accept shouldn't return an error, but we can close the fd we're doing an 
- Some I/O related fixes (make the hyperstart->shim path more robust to
  unexpected data, had forgotten to "make" one map, ..)
- Restrict the duration of each unit test to 2s to prevent 10mins timeouts on
  tests that would fail on deadlocks
- Moved the fd passing facilities to the api package, it's part of the proxy API after all
- Fix the AllocateIo client entry point to return a file corresponding to the I/O fd
- Add missing error check in the fd passing code

dlespiau commented:
Last push was just a minor layouting fix in proxy/README.md

jodh-intel commented on proxy/README.md:
```diff
@@ -0,0 +1,113 @@
+    - Arbitrate access to the `hyperstart` control channel between all the
+      instances of the OCI runtimes and `cc-shim`.
+    - Route the I/O streams between the various shim instances and `hyperstart`.
+- There's only one `cc-proxy` per host.
```
Ah - here it is. Maybe move this sentence up to the summary.

jodh-intel commented on proxy/README.md:
```diff
@@ -0,0 +1,113 @@
+a small RPC protocol, requests being similar to a function call and responses
+encoding the result of the call.
+
+Each message is composed of a header and some data:
```
Some *optional* data?

jodh-intel commented on proxy/README.md:
```diff
@@ -0,0 +1,113 @@
+Unsurprisingly, the response has the result of a command, with `success`
+indicating if the request has succeeded for not. If `success` is `true`, the
+response can carry additional return values in `data`. If success if `false`,
+`error` may contain an error string.
```
That's not very precise - can we not guarantee that *some* sort of error message will be generated? If so, that would avoid a check on whether `data` contains anything.

dlespiau replied:
It's quite precise :) but it's not guaranteeing a nice property. FWIW, today all `"success": false`  responses have an associated error message, so we can say that we guarantee the above. That said, it could be a good idea for client code to handle the case of no "error" key in the response anyway?

Still, will change it.

jodh-intel commented on proxy/api/fdpassing.go:
```diff
@@ -0,0 +1,71 @@
+	if oobn == 0 {
+		return -1, errors.New("no out of band data read")
+	}
+	if n != 1 && buf[0] != 'F' {
```
Any way to re-use `fileTag` here to avoid hard-coding `F`?

jodh-intel commented on proxy/vm.go:
```diff
@@ -0,0 +1,460 @@
+	//   hyper.INIT_WRITEFILE
+	//   hyper.INIT_READFILE
+	switch cmd {
+	case "version":
```
I almost wonder if this switch could be auto-generated?

dlespiau replied:
Maybe, but there are commands we can't have in there as we can't wrap them in JSON (writefile/readfile), so the autogeneration would have to know that. Seems like a lot of work for very little gain.

jodh-intel commented on tests/mock/hyperstart.go:
```diff
@@ -0,0 +1,377 @@
+func (h *Hyperstart) ackData(nBytes int) {
+	data := make([]byte, 4)
+	binary.BigEndian.PutUint32(data[:], uint32(nBytes))
+	//h.logf("ctl: acking %d bytes\n", nBytes)
```
Commented-out code.

jodh-intel commented on tests/mock/hyperstart.go:
```diff
@@ -0,0 +1,377 @@
+	//   hyper.INIT_WRITEFILE
+	//   hyper.INIT_READFILE
+	switch cmd {
+	case hyper.INIT_VERSION:
```
Likewise - autogenerate?

jodh-intel commented on tests/mock/hyperstart.go:
```diff
@@ -0,0 +1,377 @@
+// io channel
+//
+
+const ioHeaderSize = 12
```
Huh?

jodh-intel commented on Makefile.am:
```diff
@@ -199,6 +201,47 @@ cc_oci_runtime_CFLAGS = \
+CHECK_DEPS += check-proxy
+
+check-proxy:
+	go test -timeout 2s $(srcdir)/proxy
```
Thinking of the Travis vector, maybe run as `go test -v ...` (and possibly redirect to a file for display on failure)?

jodh-intel commented:
This looks good, although I noticed that not all the `.go` files are included in `make dist` (for example, `REAMDE.md`, `COPYING`, `*_test.go`).

Also, please can you add function headers for all go functions/receivers.

Related to this, I'd like to see the build run extra checks now that `go` source is being added. Specifically,

- `go vet ./...`
- `go fmt ./...`
- `golint ./...`

The lint tool would have caught the missing function documentation. We'll need `configure.ac` logic to check if these are available. Might be easier to just add a shell script that calls all of the above (but only warns if tools are not available?) and get the build to call that script at build time. If any available tool is run and fails, that would fail the build.

The go tools are so well integrated, I think we should start as we mean to go on and use as many as make sense for us.

As you've added a vendor directory, we should consider adding support for a dependency-tracking tool like https://launchpad.net/godeps or https://github.com/tools/godep and store the ASCII data file in git.

dlespiau commented on proxy/README.md:
```diff
@@ -0,0 +1,113 @@
+[`hyperstart`](https://github.com/hyperhq/hyperstart) VM agent to multiple
+clients on the host.
+
+![High-level Architecture Diagram](../documentation/high-level-overview.png)
```
1/ Might be worth showing that, but would rather spend more time on this right now. If you think it's worth the time, could you file an issue so we don't lose the idea?

2/ The diagram was hand drawn using google docs, not generated.

dlespiau commented:
PR updated:
  - bump the size field to 32 bits (from 16 bits) because 64k isn't enough to for exec() limits (see "Limits on size of arguments and environment" on http://man7.org/linux/man-pages/man2/execve.2.html)

dlespiau commented:
@jodh-intel: For the documentation + various linters point. We, of course, want them all. However I'd like to merge this sooner rather than later, working on the above as we go along. I'll file issues so we don't forget them. As for the "where", I'd just put them in .travis.yaml or make check (not build time).

For the vendor directory, I've actually looked at what ciao is doing and they have a vendoring tool that has some nice properties. The information is stored in https://github.com/01org/cc-oci-runtime/blob/wip/hyperstart/vendor/vendor.json. Unfortunately I've populated that data manually because the ciao vendor tool isn't quite ready to be used outside of its repo and I took the lazy approach.

dlespiau commented on proxy/README.md:
```diff
@@ -0,0 +1,113 @@
+## Payloads
+
+Payloads are in their own package and [documented there](
+https://godoc.org/github.com/01org/cc-oci-runtime/proxy/api)
```
It should be a valid one when the code is pushed to master and that package path exists! hopefully.

gorozco1 commented:
lgtm

devimc commented:
:+1:

chavafg commented:
qa-failed

chavafg commented:
only make distcheck failed, moving to qa-passed

chavafg commented:
qa-passed
